### ISSUE TITLE
PyLance skipping UNC paths "because it is not a valid directory"

### ISSUE THREAD

**Issue by vaughajm** [microsoft/pylance-release#4357]
<h3> Environment data </h3>

-   Language Server version: 2023.5.20
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: off

<h3> Code Snippet </h3>

```python
import pyodbc
print('ImaTest')

print(pyodbc.drivers())
```

<h3> Repro Steps </h3>

1. XXX

<h3> Expected behavior </h3>

Should show no errors as pyodbc is installed at:

\\jcs1\sqltransfer\Python\Python310\Lib\site-packages

And the interpreter is pointing to:

\\jcs1\sqltransfer\Python\Python310\Lib\python.exe

(This worked recently.)

<h3> Actual behavior </h3>

pyodbc is highlighted and PyLance is reporting:

Import "pyodbc" could not be resolved Pylance[reportMissingImports]

<h3> Logs </h3>

```
XXX
```
[PyLance.log](https://github.com/microsoft/pylance-release/files/11466943/PyLance.log)

**Comment by rchiodo** [comment]
Thanks I can repro the issue.

Got this after playing around with WSL:

```
2023-05-12 13:29:05.650 [info] [Info  - 1:29:05 PM] (25508)   Skipping '\\wsl.localhost\Ubuntu\home\rich\windows_venv' because it is not a valid directory
```

The code is here:

https://github.com/microsoft/pyright/blob/b7f2751b3a7ebdb358c0330f87d23b46f0827513/packages/pyright-internal/src/common/fullAccessHost.ts#L222

I don't believe this has changed in the last two years though. That eventually goes down to here to detect if the file exists:
https://github.com/microsoft/pyright/blame/b7f2751b3a7ebdb358c0330f87d23b46f0827513/packages/pyright-internal/src/common/realFileSystem.ts#L213

**Comment by rchiodo** [comment]
Maybe we updated yarnfs recently?

**Comment by rchiodo** [comment]
Yeah looks like dependabot updates brought it in. 
https://github.com/microsoft/pyright/blame/b7f2751b3a7ebdb358c0330f87d23b46f0827513/packages/pyright-internal/package.json#L19

which ends up here:
https://github.com/microsoft/pyrx/blame/57bb0ddf8408602820d1f9adf83b68cb88cf8a3f/packages/pyright/packages/pyright-internal/package.json#L19

So my guess is yarnfs regressed something (or changed the way we should check for existence)

**Comment by blerrgh** [comment]
I have the same issue on windows 11, even with a network drive mapped to a drive letter (Z:\\). 

I don't get a "is not a valid directory" error, and this is in the logs:
```
[info] [Info  - 5:59:51 PM] (24120) Setting pythonPath for service "test_project": "z:\test_project\.venv\Scripts\python.exe"
```

But I'm still receiving erroneous `Import could not be resolved` errors.

This only started recently, in the last few days, but I'm not exactly what triggered it.

**Comment by blerrgh** [comment]
This is almost definitely caused by https://github.com/microsoft/vscode/issues/182070, but following the instructions here https://code.visualstudio.com/docs/setup/windows#_working-with-unc-paths , neither using a mapped drive nor adding my network server to `security.allowedUNCHosts` resolved the issue for me.

**Comment by philnagel** [comment]
Some observations related to this as I am also experiencing this issue. I just learned about this new UNC Path whitelist settings, and even after allowing my specific host, I am still seeing this issue.

- When setting the log level to `Trace`, I do see entries that read `Skipping '\\hostname\[...]' because it is not a valid directory` even when the directory exists. 
- This skipping message does not appear for a UNC path that is included in `python.analysis.extraPaths`, but it does appear for a UNC path that is included in my Python environment via a `.pth` file (both in the same host).
- For UNC paths included in `python.analysis.extraPaths` I see:
`Looking in extraPath \\hostname\[...]`
`Attempting to resolve stub package using root path \\hostname\[...]`
`Attempting to resolve using root path \\hostname\[...]`
but no other feedback
- Using a mapped drive pointed to a UNC path also does not work. I am still seeing the "Import could not be resolved" message.
- using an older version of VS Code does seem to resolve the above problem.

**Comment by blerrgh** [comment]
Looks like they're implementing a few changes to VS Code that may or may not affect this Pylance issue:

https://github.com/microsoft/vscode/issues/182070#issuecomment-1567199832

I haven't had a chance to try an Insider build with these changes to see if it resolves this Pylance issue or not.  Hopefully a proper VS Code release will be along soon.

**Comment by philnagel** [comment]
Since the `security.allowedUNCHosts` does not seem to affect this behavior, I am wondering if there is something additional going on on top of the general VS Code changes.

**Comment by blerrgh** [comment]
Updating to the latest VS Code release resolved all these issues for me.

https://github.com/microsoft/vscode/releases/tag/1.79.0

**Comment by vaughajm** [comment]
Still seeing exactly the same issue.

**Comment by philnagel** [comment]
I just downloaded 1.79.0 and I am still seeing this issue. For what it's worth, this is when using the Remote SSH extension from macOS to a Windows machine.
I did set the new setting `"security.restrictUNCAccess": true` (by the way, this setting along with `security.allowedUNCHosts` have disappeared from the settings UI in this new version). I am still getting the message `Skipping [uncpath]` in the output of the Python Language Server. Another observation I just made (not sure if relevant) is that the `Skipping` message only shows if the UNC path is in the Python path of the interpreter (in my case I added a `.pth` file). When the path is found only in `python.analysis.extraPaths`, the message does not show but the import is still not resolved either.

**Comment by philnagel** [comment]
@rchiodo are there any updates on this or any further info we can provide to help fix this?

**Comment by rchiodo** [comment]
I don't believe this is a pylance issue. It should be a VS code issue/feature as they disabled UNC paths on purpose. See the VS code issues linked above:
https://github.com/microsoft/vscode/issues/182070
https://github.com/microsoft/vscode/issues/182070#issuecomment-1567199832

You might want to try with VS code insiders instead.

**Comment by philnagel** [comment]
I tried both 1.79.2 and Insider's and I'm still seeing the issue. At least Intellisense works now, but the import is still being flagged as unresolved. I [opened an issue](https://github.com/microsoft/vscode/issues/185807) with VS Code.

**Comment by bpasero** [comment]
Fyi the changes VS Code did to block UNC will only impact an extension if the extension is using node.js or `vscode.workspace.fs` for file access, but for example not when any other language is used, such as Python.

**Comment by bpasero** [comment]
@rchiodo there is something strange going on only with the PyLance extension. In an experiment in https://github.com/microsoft/vscode/issues/185807#issuecomment-1611783239, we asked the user to see if the TypeScript extension also sees the UNC path blocked and it turns out that it is not blocked for the TypeScript extensions.

So please shed some light how your extension works, i.e. how you access files that lead to the issue. Are you maybe forking a process yourself as a child to do it? Are you using node.js or our API for file access?

**Comment by rchiodo** [comment]
The access that's erroring is in a subprocess of the extension process. It's a node.js process that is running the language server. 

The code for that is actually open source. It's here:
https://github.com/microsoft/pyright/blob/74d8f3c4d273f23a1f6677510e76d2efd11222b3/packages/pyright-internal/src/common/realFileSystem.ts#L216

Pylance is derived from the open source Pyright language server. 

Under the covers that is using yarnfs, which looks like it just uses the node fs.
https://github.com/yarnpkg/berry/blob/dbfb5007fae4aa639fa969536a9158d391b347c7/packages/yarnpkg-fslib/sources/NodeFS.ts#L147

**Comment by rchiodo** [comment]
Does the UNC workaround only work in extension host?

**Comment by bpasero** [comment]
@rchiodo yes, the UNC allow-list / block-list works in the extension host and should work in all sub-processes forked from there too. 

> The access that's erroring is in a subprocess of the extension process.

Can you share the code how this subprocess is being created?

**Comment by rchiodo** [comment]
It's created using the vscode-languageclient. I believe it ends up at this code here:

https://github.com/microsoft/vscode-languageserver-node/blob/2041784436fed53f4e77267a49396bca22a7aacf/client/src/node/main.ts#L353-L354

So spawn instead of fork.

**Comment by rchiodo** [comment]
Nope sorry I was wrong. It's going down the fork path. Here:
https://github.com/microsoft/vscode-languageserver-node/blob/2041784436fed53f4e77267a49396bca22a7aacf/client/src/node/main.ts#L396

**Comment by bpasero** [comment]
@rchiodo where is the sources of your extension for us to maybe debug and verify?

**Comment by rchiodo** [comment]
https://github.com/microsoft/pyrx. Hold on I'll give you access.

Although you could just use pyright, it should have the same problem as its logic is the same:
https://github.com/microsoft/pyright

**Comment by vaughajm** [comment]
Seems to be working fine now.

**Comment by rchiodo** [comment]
@bpasero I've finally been able to reproduce this with a simple extension. I added a new bug for it: 

https://github.com/microsoft/vscode/issues/194662
